geoutils/filters.py

```diff
@@ -0,0 +1,296 @@
+
+    filtered[mask] = np.nan
+
+    return filtered
+
+
+@jit(nopython=True, parallel=True)  # type: ignore
+def median_filter_numba(array: NDArrayNum, window_size: int) -> NDArrayNum:
+    """
+    Apply a median filter to a raster that may contain NaNs, using numbas's implementation.
+
+    :param array: The input array to be filtered.
+    :param window_size: The size of the window to use (must be odd).
+
+    :returns: The filtered array (same shape as input).
+    """
+
+    if window_size % 2 == 0:
+        raise ValueError("window_size must be odd")
+
+    # Get input shapes
+    N1, N2 = array.shape
+
+    # Define ranges to loop through given padding
+    row_range = N1 - window_size + 1
+    col_range = N2 - window_size + 1
+
+    # Allocate an output array
+    outputs = np.full((row_range, col_range), fill_value=np.nan, dtype=np.float32)
+
+    # Loop over every pixel concurrently by using prange
+    for row in prange(row_range):
+        for col in prange(col_range):
+
+            outputs[row, col] = np.nanmedian(array[row : row + window_size, col : col + window_size])
+
+    return outputs
+
+
+def median_filter(array: NDArrayNum, window_size: int, engine: Literal["scipy", "numba"] = "numba") -> NDArrayNum:
+    """
+    Apply a median filter to a raster that may contain NaNs, using scipy's implementation.
+
+    :param array: The input array to be filtered.
+    :param window_size: The size of the window to use (must be odd).
+    :param engine: Filtering engine to use, either "scipy" or "numba".
+
+    :returns: The filtered array (same shape as input).
+    """
+
+    if window_size % 2 == 0:
+        raise ValueError("window_size must be odd")
+
+    if array.ndim == 2:
+        return _apply_median_filter_2d(array, window_size, engine)
+    elif array.ndim == 3:
+        return np.stack([_apply_median_filter_2d(slice_, window_size, engine) for slice_ in array])
+    else:
+        raise ValueError("Input array must be 2D or 3D.")
+
+
+def _apply_median_filter_2d(array: NDArrayNum, window_size: int, engine: Literal["scipy", "numba"]) -> NDArrayNum:
+    """
+    Apply a 2D median filter on an array that may contain NaNs.
+
+    :param array: 2D input array to filter, may contain NaNs.
+    :param window_size: Size of the median filter window (must be odd).
+    :param engine: Filtering engine to use, either "scipy" or "numba".
+    :returns: Filtered array of the same shape as input.
+    """
+
+    if engine == "scipy":
+        return scipy_generic_filter(array, np.nanmedian, size=window_size, mode="constant", cval=np.nan)
+    else:
+        if not _has_numba:
+            raise ValueError("Optional dependency needed. Install 'numba'.")
+        hw = int((window_size - 1) / 2)
+        array = np.pad(array, pad_width=((hw, hw), (hw, hw)), constant_values=np.nan)
+        return median_filter_numba(array, window_size)
+
+
+def mean_filter(array: NDArrayNum, kernel_size: int = 5, **kwargs: Any) -> NDArrayNum:
```

Comment:
is there any reason to set a default value here? I would not make 5 the default but always force the user to set it.